### PR TITLE
[FSDP] Move the sharded_state_dict logic to the post hook to avoid OOM

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82613

The original implementation put the call of `_summon_full_params()` in `state_dict()`. However, because `state_dict()` is recursive, `_summon_full_params()` will also behave like the recursive version even if recursive is set to False. This PR put the logic in the post hook to solve the OOM issue.

Differential Revision: [D38329396](https://our.internmc.facebook.com/intern/diff/D38329396/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1983,19 +1983,25 @@ def _sharded_post_state_dict_hook(
...
+                offsets = [0 for _ in param.size()]
+                offsets[0] = math.ceil(param.size()[0] / self.world_size) * self.rank
+                local_shards = [
+                    Shard.from_tensor_and_offsets(local_shard, offsets, self.rank)
+                ]
+                fqn = f"{prefix}{fqn}"
+                state_dict[fqn] = init_from_local_shards(
+                    local_shards, param.size(), process_group=self.process_group
+                )  # type: ignore[assignment]
+        state_dict.pop(f"{prefix}{FLAT_PARAM}")
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2004]: I'm assuming that this is removing the key checkpointed by the `super().state_dict()` call?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -2103,24 +2109,19 @@ def state_dict(self, *args, **kwargs):
             else:
                 return {}
 
-        elif self._state_dict_type == StateDictType.LOCAL_STATE_DICT:
+        elif (
+            self._state_dict_type == StateDictType.LOCAL_STATE_DICT or
+            self._state_dict_type == StateDictType.SHARDED_STATE_DICT
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2114]: so it seems that sharded state dict calling `state_dict` is not meant to do much, we remove the checkpointed FLAT_PARAM and add new data to the state_dict for the sharded original parameters. 

If this is the case, could we just remove the super().state_dict calls, recurse ourselves and call the post hook?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1983,19 +1983,25 @@ def _sharded_post_state_dict_hook(
...
-            ]
-            state_dict[fqn] = init_from_local_shards(
-                local_shards, param.size(), process_group=self.process_group
-            )  # type: ignore[assignment]
+        assert self.training_state != TrainingState_.SUMMON_FULL_PARAMS, (
+            "Inside _sharded_post_load_state_dict_hook, the training_state must "
+            "not be SUMMON_FULL_PARAMS."
+        )
+        with self._summon_full_params(recurse=False, writeback=False):
+            for fqn, _, _ in self._param_fqns:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1991]: wouldn't `named_parameters()` also just give the parameter names we need?

**Review by rohan-varma — APPROVED** [review]
Thanks for the fix! It would be great to test it on a use case where full state dict fails but this one succeeds.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2004]: Yes

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1991]: named_parameters() will give us more parameters than what we need. We need to use recursive named_parameters() and this will give us more parameters: 1.) parameters in children FSDP modules 2.) parameters that are ignored by FSDP.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2114]: We need the recursive calls for 1.) constructing the correct prefix, 2.) calling post_hook in a reversed order. We can do this ourselves but it will be better to reuse state_dict logic. The only thing that sharded_state_dict does not need is the detach one but that should not causing too many overheads.
